### PR TITLE
segmentation.watershed returns wrong data type bug fix

### LINKED ISSUE #6587: `segmentation.watershed` returns wrong data type
### Description:

The documentation of `segmentation.watershed` says that:

> ### Returns
>   **out**: ndarray
>   A labeled matrix of the same type and shape as markers

[[0.18.x]](https://scikit-image.org/docs/0.18.x/api/skimage.segmentation.html#skimage.segmentation.watershed) [[0.19.x]](https://scikit-image.org/docs/0.19.x/api/skimage.segmentation.html#skimage.segmentation.watershed)

I have tested this with 0.18.1:

```python
import skimage.segmentation
import numpy as np
print(skimage.segmentation.watershed(np.zeros((100, 100)), np.zeros((100, 100), np.uint16)).dtype)
```

Gives `int32` but `uint16` is expected.

### Way to reproduce:

```Python
import skimage.segmentation
import numpy as np
print(skimage.segmentation.watershed(np.zeros((100, 100)), np.zeros((100, 100), np.uint16)).dtype)
```

### Traceback or output:

```Shell
int32
```

### Version information:

```Shell
3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0]
Linux-5.8.0-36-generic-x86_64-with-glibc2.10
scikit-image version: 0.18.1
numpy version: 1.20.3
```

### PR DESCRIPTION
**Pull request by decorouz**
## Description
Bug fix: segmentation.watershed returns wrong data type #6587

(Fixes #6587)

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

File: skimage/segmentation/_watershed.py
Comment (jni) [inline skimage/segmentation/_watershed.py:29]: Please revert this change — see [PEP-257](https://peps.python.org/pep-0257/) for docstring guidelines.

File: skimage/segmentation/_watershed.py
Comment (jni) [inline skimage/segmentation/_watershed.py:122]: Ideally, we should change the function so that the same dtype as the markers is returned, rather than the documentation. It shouldn't be too hard a change.

**Review by jni — CHANGES_REQUESTED** [review]
@decorouz I know @stefanv suggested that changing the docs would be ok, but I think it's worth actually fixing the bug rather than changing the documentation... However, that might involve changing some of the Cython code to use [anyint](https://github.com/scikit-image/scikit-image/blob/907bfb71d2c8e4d58c2dffc6056760c1e00d070b/skimage/_shared/fused_numerics.pxd#L16-L18) rather than [DTYPE_INT32](https://github.com/scikit-image/scikit-image/blob/907bfb71d2c8e4d58c2dffc6056760c1e00d070b/skimage/segmentation/_watershed_cy.pyx#L38). You can see how to import fused numeric types in [this file](https://github.com/scikit-image/scikit-image/blob/49cb3a4d4c17f1c88e86aa86401066234ad441f7/skimage/util/_remap.pyx#L3), for example.

File: skimage/segmentation/_watershed.py
Comment (decorouz) [reply skimage/segmentation/_watershed.py:29]: I did attempt to fix the bug the ideal way as suggested by @stefanv. However, I got a `ValueError` from one of the Cypthon files. Thanks for pointing me to a resources that should help.
I will revert the changes and attempt the fix the bug appropiately

File: skimage/segmentation/_watershed.py
Comment (decorouz) [reply skimage/segmentation/_watershed.py:122]: Hi @jni , When I test my approach locally, I get the following error after changing every instance of `'DTYPE_INT32_t'` to `np_anyint` in the `_watershed_cy.pyx` file. 

`ValueError: Buffer dtype mismatch, expected 'DTYPE_INT32_t' but got 'unsigned short'`

All test passed okay. Any suggestions ?

Comment (jni) [reply skimage/segmentation/_watershed.py:122]: Are you sure you've changed *every* instance? It sounds from the error like you missed one. (There is more than one relevant file, also heap_watershed, for example) If you push your changes I can have a look.

The tests pass probably because the compilation failed but the old compiled functions are still there.

Comment (decorouz) [reply skimage/segmentation/_watershed.py:122]: I did change all the instances in my previous attempts. I figured I was testing my changes wrong hence the `ValueError`. Eureka

File compiled successfully and test passed. 
I have pushed my changes. Please review.
Thank you.

File: skimage/segmentation/_watershed.py
Comment (jni) [inline skimage/segmentation/_watershed.py:82]: You can just remove "astype" here:

```suggestion
            markers,
```

File: skimage/segmentation/_watershed.py
Comment (jni) [reply skimage/segmentation/_watershed.py:29]: Don't forget to revert this change!

**Review by jni — CHANGES_REQUESTED** [review]
@decorouz this is fantastic! 🤩 

I'd like to request three further changes:

1. Could you please add a test to check that passing markers of different dtypes (uint8, int8, uint16, int16, uint32, int32, uint64, int64) results in a watershed result of that dtype? You can look at [this file](https://github.com/scikit-image/scikit-image/blob/main/skimage/segmentation/tests/test_watershed.py) for test examples (and that is where the new test should go).
2. Could you revert your original documentation change specifying that the output is int32, now that the output is whatever the input was? 😊 
3. In my suggested correction about not using `markers.astype()` but just `markers`, I haven't accounted for the possibility that the input markers array would get overwritten, which would be bad. So, could you add another test to show that the markers array is not modified? Something like:

```python
makers_original = np.copy(markers)
result = watershed(...)
np.testing.assert_equal(makers_original, markers)
```

Thanks! 🚀

File: skimage/segmentation/_watershed.py
Comment (decorouz) [reply skimage/segmentation/_watershed.py:29]: Sure.

**Comment by decorouz** [comment]
> @decorouz this is fantastic! 🤩
> 
> I'd like to request three further changes:
> 
> 1. Could you please add a test to check that passing markers of different dtypes (uint8, int8, uint16, int16, uint32, int32, uint64, int64) results in a watershed result of that dtype? You can look at [this file](https://github.com/scikit-image/scikit-image/blob/main/skimage/segmentation/tests/test_watershed.py) for test examples (and that is where the new test should go).
> 2. Could you revert your original documentation change specifying that the output is int32, now that the output is whatever the input was? 😊
> 3. In my suggested correction about not using `markers.astype()` but just `markers`, I haven't accounted for the possibility that the input markers array would get overwritten, which would be bad. So, could you add another test to show that the markers array is not modified? Something like:
> 
> ```python
> makers_original = np.copy(markers)
> result = watershed(...)
> np.testing.assert_equal(makers_original, markers)
> ```
> 
> Thanks! 🚀

I have implemented the changes in my recent push. Please review. 
Thanks @jni

File: skimage/segmentation/tests/test_watershed.py
Comment (jni) [inline skimage/segmentation/tests/test_watershed.py:404]: Ah, a couple of comments on this test @decorouz:

1. Please rename the test, `test_watershed_input_not_modified`. (I see that there's other poorly-named tests above it, but today is the best day to start improving our test names! 😃)
2. Unfortunately, in this situation the test is not going to guarantee correctness, because there are no markers (markers have value >0), so watershed will do nothing. I suggest instead

```python
image = np.random.default_rng().random(size=(21, 21))
markers = np.zeros((21, 21), dtype=np.uint8)
markers[[5, 5, 15, 15], [5, 15, 5, 15]] = [1, 2, 3, 4]
original_markers = np.copy(markers)
result = watershed(image, markers)
np.testing.assert_equal(original_markers, markers)
assert not np.all(result == markers)
```

The last guard will prevent the issue I raised above from reappearing.

I think this is really the last thing! 😅 Thank you! 🙏

Comment (jni) [reply skimage/segmentation/tests/test_watershed.py:404]: (Note: in case you're unfamiliar with NumPy array fancy indexing on line 3:

https://numpy.org/doc/stable/user/basics.indexing.html#integer-array-indexing
)

Comment (decorouz) [reply skimage/segmentation/tests/test_watershed.py:404]: I did consider naming the test differently, however since the other tests in the class followed a similar pattern, I decided to follow the same. Your suggestion is better. And yes, It's test names improvement day 😃. 

Thank you for your reviews and contributions. I hope it is the last thing for this issue. I enjoyed it.

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
Great stuff @decorouz!!! It is impressive to play with Cython as a first contribution! 😅 Typically we leave PRs sitting for around ~24h before merging in case someone else from the core team has more comments. Feel free to ping again if it takes much longer than that! Thank you! 😊

File: skimage/segmentation/tests/test_watershed.py
Comment (mkcor) [inline skimage/segmentation/tests/test_watershed.py:443]:
```suggestion
    'dtype', [np.uint8, np.int8, np.uint16, np.int16,
              np.uint32, np.int32, np.uint64, np.int64]
```
I would think...?

File: skimage/segmentation/tests/test_watershed.py
Comment (mkcor) [inline skimage/segmentation/tests/test_watershed.py:399]:
```suggestion
        """Test to ensure input markers are not modified."""
```

**Review by mkcor — COMMENTED** [review]
Nice work, @decorouz! I have just left a couple of suggestions.

File: skimage/segmentation/tests/test_watershed.py
Comment (mkcor) [reply skimage/segmentation/tests/test_watershed.py:399]: Hmm, maybe my suggestion is not ideal, because it sounds like "Test" is a noun, not a verb... Otherwise, "Test that input markers are indeed not modified." :wink:

**Comment by decorouz** [comment]
> Nice work, @decorouz! I have just left a couple of suggestions.

Thank you. Appreciate the suggestions.

**Comment by jni** [comment]
By the way, the failing tests now are because of jupyter-widgets/ipywidgets#3731, unrelated to this PR.

**Comment by decorouz** [comment]
> By the way, the failing tests now are because of [jupyter-widgets/ipywidgets#3731](https://github.com/jupyter-widgets/ipywidgets/issues/3731), unrelated to this PR.

I noticed. Thanks for the heads up.

File: skimage/segmentation/tests/test_watershed.py
Comment (lagru) [inline skimage/segmentation/tests/test_watershed.py:446]:
```suggestion
    assert out.dtype == markers.dtype
```

**Review by lagru — APPROVED** [review]
Looks good @decorouz. I can only agree with @jni, good work for what I'm assuming to be your first time with Cython. :) I'll merge in a bit but want to make sure that the CI isn't complaining about my suggestion.

**Pull request merged**
